_posts/2022-05-23-fine-grained-access-control-for-multitenancy.markdown

```diff
@@ -0,0 +1,326 @@
+
+![image](architecture.png)
+
+1. Users authenticate with a username and password, the web app passes these to amazon cognito for validation.
+2. If the supplied credentials (username and password) are valid, cognito creates a session and subsequently issues three (3) JWTs (JSON Web Tokens). The aforementioned tokens are id token, access token and a refresh token. The authenticated user can now send requests to api gateway along with with the id token in the headers section.
+3. API gateway sends the received id token to a lambda function called an authorizer.
+4. The authorizer function verifies the claims attached to the id token.
+5. The authorizer returns a policy and context.
+6. API Gateway evaluates the policy and forwards the request to a lambda function along with the authorizer generated context.
+7. The lambda function writes/reads data according to the tenantId listed in the forwarded context.
+8. A response is returned by the lambda function.
+
+### Tenant Isolation
+
+#### Tenant ID
+
+To create a secure `multi-tenant` environment there needs to be some notion of `tenant resource isolation`. In essence, `tenant A` should not be able to access the resources of `tenant B` and vice versa. To solve this problem it was decided that it is necessary to assign a unique identifier called `tenantID` to each user/customer. The aforementioned `tenantID` is a composite string that is generated and then attached to the calling user as a `custom attribute` during the registration stage.
+
+After successful authentication the `tenantID` custom attribute becomes available as a parameter within the `ID token` as a parameter with the following key `custom:tenantID`. It is important to note that access tokens do not carry any of the user's custom attributes, only id tokens have this capability. In saying so, the application will exclusively use id tokens for authorization/authentication purposes.
+
+#### Lambda Context Objects
+
+In order to make use of the idea of a `tenantID` it is important to realise that there needs to be some way of retrieving the correct `tenantID` parameter and propagating it to the running lambda function's execution scope. This is where the idea of Lambda context objects comes into play.
+
+Context objects can be modified to include custom parameters that can then be accessed during run time. With this in mind, it should be relatively straight forward to use the application's Lambda Authorizer to forward the retrieved `tenantID` as a custom parameter within the JSON output of the lambda authorizer.
+
+When downstream lambdas are invoked they cam access the context object as a key within the event object. Take for example the following sample lambda event object.
+
+```
+2022-05-20T08:09:58.116Z	1c1c55e3-6ec1-47fb-962f-d94387e10480	INFO	{
```

Comment:
Suggestion; It might be worth pointing out to the reader that it's not generally a good idea to print sensitive information out to logs.